# TP NoSQL

Prénom :

Nom : 

Pour ce TP noté vous avez à trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [65]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [66]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://mongo_user:wb6xlTxueSQ9KU00@cluster-but-sd.8m2t0.mongodb.net/?retryWrites=true&w=majority&appName=cluster-but-sd'
client = pymongo.MongoClient(URI)
db = client.tp
data = db.airbnb
# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['restaurants', 'airbnb']


#### 1. Combien de logements sont la base de données ?

In [110]:
result = data.count_documents({})
print("il y a",result,"logements dans la base de données.")

il y a 5555 logements dans la base de données.


#### 2. Quel est le prix moyen par ville ? Trier les villes par prix décroissant.

In [109]:
result = data.aggregate([
    {"$group":{"_id":"$address.market","Prix moyen":{"$avg":"$price"}}}
])
pd.DataFrame(result)

_id                           Prix moyen
0                    Oahu  212.2964426877470355731225296442688
1          The Big Island  179.4604316546762589928057553956835
2                   Kauai  288.8208955223880597014925373134328
3                New York  139.6260296540362438220757825370675
4                Istanbul  367.9454545454545454545454545454545
5                    Maui  286.5882352941176470588235294117647
6        Other (Domestic)                               128.00
7                                                       115.50
8                  Sydney  197.7142857142857142857142857142857
9                   Porto  69.12815884476534296028880866425993
10              Barcelona  100.9477848101265822784810126582278
11               Montreal  100.2330246913580246913580246913580
12              Hong Kong  762.4781906300484652665589660743134
13         Rio De Janeiro  525.8059701492537313432835820895522
14  Other (International)                               445.75

Le tableau ci-dessus contient le prix moyen des logements par ville.
Par exemple, le prix moyen à New York est 139.63 Dollards.


#### 3. Afficher la liste de tous les différents équipements qui existent.

In [104]:
result = data.distinct("amenities")
pd.DataFrame(result)

0
0                                              
1                              24-hour check-in
2                         Accessible-height bed
3                      Accessible-height toilet
4                              Air conditioning
..                                          ...
181                                        Wifi
182                               Window guards
183                                      toilet
184  translation missing: en.hosting_amenity_49
185  translation missing: en.hosting_amenity_50

[186 rows x 1 columns]

La liste ci-dessus montre les 185 équipements différents.

#### 4. Combien de propriétés incluent le Wifi dans les équipements ?

In [108]:
result = data.count_documents({ "amenities": "Wifi" })
print("Il y a ",result," logements dans lesquels le Wifi est inclu.")

Il y a  5303  logements dans lesquels le Wifi est inclu.


#### 5. Afficher le nom de tous les logements ainsi que le nombre de chambres et de lits qu'ils contiennent (ne pas afficher l'ID)

In [139]:
result = data.aggregate([
     {"$project":{"_id":0,"name":1,"bedrooms":1,"beds":1}}
 ])
pd.DataFrame(result)

name  bedrooms  beds
0                          Apt Linda Vista Lagoa - Rio       1.0   1.0
1                   Ótimo Apto proximo Parque Olimpico       2.0   2.0
2                       Modern Spacious 1 Bedroom Loft       1.0   2.0
3     Ligne verte - à 15 min de métro du centre ville.       0.0   1.0
4                              Ribeira Charming Duplex       3.0   5.0
...                                                ...       ...   ...
5550     Estudio en el casco antiguo. Alquiler mensual       0.0   3.0
5551  Qto Aconchegante no em frente ao Parque Olímpico       1.0   1.0
5552                  StayIN Oporto Música - Apartment       8.0   8.0
5553                                      Quarto Legal       1.0   1.0
5554                                                         1.0   1.0

[5555 rows x 3 columns]

Le tableau ci-dessus montre la liste des logements avec le nombre de chambre et de lit

Par exemple, Quarto Legal a 1 chambre et 1 lit

#### 6. Afficher le nom et le prix des logements situés à Porto.

In [144]:
result = data.aggregate([
    { "$match": { "address.market": "Porto" }},
    {"$project":{"_id":0,"name":1,"price":1}}
 ])
pd.DataFrame(result)

name   price
0                Ribeira Charming Duplex   80.00
1                      Be Happy in Porto   30.00
2    Downtown Oporto Inn (room cleaning)   40.00
3                         FloresRooms 3T   31.00
4     A Casa Alegre é um apartamento T1.   40.00
..                                   ...     ...
549         Number283 - Your home abroad   48.00
550        Cozy apartment downtown Porto   58.00
551                   Soares Guest House   90.00
552   Central, Modern, University, Metro   41.00
553     StayIN Oporto Música - Apartment  500.00

[554 rows x 2 columns]

Le tableau ci-dessus comprend tout les logements proposé à porto ainsi que leur prix.

Par exemple, FloresRooms 3T est disponible pour 31 euros

#### 7. Quels sont les 5 hôtes les plus populaires (ceux dont les propriétés ont reçu le plus de commentaires) ?

In [ ]:
result = data.aggregate([
    {}
])

#### 8. Quelles sont les 6 villes ayant le plus de logements disponibles à la location ?

#### 9. Combien de propriétés acceptent plus de 4 invités et ont une caution de moins de 300€ ?

#### 10. Donner les 20 utilisateurs qui ont fait le plus de commentaires (afficher seulement l'ID et le nom de l'utilisateur).

#### 11. Parmi les logements à Sydney, quel est la note moyenne des visiteurs ?

#### 12. (bonus) Afficher les logements qui contiennent le mot "park" dans leur nom

#### 13. (bonus) Afficher le nom des logements ayant une lattitude comprise entre 36,1 et 40,6 dont le prix est entre 100 et 200 euros.